src/ImageSharp/Formats/ImageDecoder.cs

```diff
@@ -0,0 +1,241 @@
+    /// </remarks>
+    /// <param name="options">The general decoder options.</param>
+    /// <param name="stream">The <see cref="Stream"/> containing image data.</param>
+    /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+    /// <returns>The <see cref="IImageInfo"/> object.</returns>
+    /// <exception cref="ImageFormatException">Thrown if the encoded image contains errors.</exception>
+    protected abstract IImageInfo Identify(DecoderOptions options, Stream stream, CancellationToken cancellationToken);
+
+    /// <summary>
+    /// Performs a resize operation against the decoded image. If the target size is not set, or the image size
+    /// already matches the target size, the image is untouched.
+    /// </summary>
+    /// <param name="options">The decoder options.</param>
+    /// <param name="image">The decoded image.</param>
+    protected static void Resize(DecoderOptions options, Image image)
+    {
+        if (ShouldResize(options, image))
+        {
+            ResizeOptions resizeOptions = new()
+            {
+                Size = options.TargetSize.Value,
+                Sampler = options.Sampler,
+                Mode = ResizeMode.Max
+            };
+
+            image.Mutate(x => x.Resize(resizeOptions));
+        }
+    }
+
+    /// <summary>
+    /// Determines whether the decoded image should be resized.
+    /// </summary>
+    /// <param name="options">The decoder options.</param>
+    /// <param name="image">The decoded image.</param>
+    /// <returns><see langword="true"/> if the image should be resized, otherwise; <see langword="false"/>.</returns>
+    private static bool ShouldResize(DecoderOptions options, Image image)
+    {
+        if (options.TargetSize is null)
+        {
+            return false;
+        }
+
+        Size targetSize = options.TargetSize.Value;
+        Size currentSize = image.Size();
+        return currentSize.Width != targetSize.Width && currentSize.Height != targetSize.Height;
+    }
+
+    internal static T WithSeekableStream<T>(
+        DecoderOptions options,
+        Stream stream,
+        Func<Stream, T> action)
+    {
+        Guard.NotNull(options, nameof(options));
+        Guard.NotNull(stream, nameof(stream));
+
+        if (!stream.CanRead)
+        {
+            throw new NotSupportedException("Cannot read from the stream.");
+        }
+
+        T Action(Stream s, long position)
+        {
+            T result = action(s);
+
+            // Our buffered reads may have left the stream in an incorrect non-zero position.
+            // Reset the position of the seekable stream if we did not read to the end to allow additional reads.
+            if (stream.CanSeek && stream.Position != s.Position && s.Position != s.Length)
+            {
+                stream.Position = position + s.Position;
+            }
```

Comment:
It's a fix for #2259 where 3 images were embedded in a stream one after the other. We have tests to cover it. I can add a reference to the issue in the code.

Reply:
I think I'm either missing something, or this code might be buggy.

If `stream.CanSeek == false`, we are creating a `ChunkedMemoryStream` and passing it to `Action`, but in that case the `stream.CanSeek && ... ` condition cannot be ever true, so we won't rewind anything, therefore `Action` is basically a dummy.

Reply:
I'll have another read through; `ChunkedMemoryStream` can seek though. 

The issue only really happens because BufferedReadStream will often overshoot the expected position since it does a bulk read. We only use that stream in our internal decoders so I should move the fix to the extension method that uses it making it much more readable and only appear one time.

Reply:
Wherever it lands, I would unit test the logic, if possible. No sure `Issue_2259.png` is enough to exercise all the cases.

Reply:
OK. There was no need for the logic in `Image.FromStream` so I've removed it. `ImageDecoder` is the only place now we perform the check which makes sense since we only use `BufferedReadStream` internally on decoders inheriting that type.